From 546050a1a032d6ae17230d6906852ee878b5aa30 Mon Sep 17 00:00:00 2001
From: zhyie <128297362+zhyie@users.noreply.github.com>
Date: Mon, 1 Jun 2026 10:17:35 +0800
Subject: [PATCH] Apply xcursor patch

---
 config.mk |  5 ++++-
 drw.c     |  5 +++--
 drw.h     |  2 +-
 dwm.c     | 10 +++++-----
 4 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/config.mk b/config.mk
index 713c949..e700319 100644
--- a/config.mk
+++ b/config.mk
@@ -58,9 +58,12 @@ XCBLIBS = -lX11-xcb -lxcb -lxcb-res
 #BDINC = `pkg-config --cflags fribidi`
 #BDLIBS = `pkg-config --libs fribidi`
 
+# Uncomment for xcursor patch
+XCURSOR = -lXcursor
+
 # includes and libs
 INCS = -I${X11INC} -I${FREETYPEINC} ${YAJLINC} ${PANGOINC} ${BDINC}
-LIBS = -L${X11LIB} -lX11 ${XINERAMALIBS} ${FREETYPELIBS}  ${XRENDER} ${MPDCLIENT} ${XEXTLIB} ${XCBLIBS} ${KVMLIB} ${PANGOLIB} ${YAJLLIBS} ${IMLIB2LIBS} $(BDLIBS) $(XILIB)
+LIBS = -L${X11LIB} -lX11 ${XINERAMALIBS} ${FREETYPELIBS}  ${XRENDER} ${MPDCLIENT} ${XEXTLIB} ${XCBLIBS} ${KVMLIB} ${PANGOLIB} ${YAJLLIBS} ${IMLIB2LIBS} $(BDLIBS) $(XILIB) $(XCURSOR)
 
 # flags
 CPPFLAGS = -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_XOPEN_SOURCE=700L -DVERSION=\"${VERSION}\" ${XINERAMAFLAGS}
diff --git a/drw.c b/drw.c
index 8a479fb..55a7379 100644
--- a/drw.c
+++ b/drw.c
@@ -4,6 +4,7 @@
 #include <string.h>
 #include <X11/Xlib.h>
 #include <X11/Xft/Xft.h>
+#include <X11/Xcursor/Xcursor.h>
 
 #include "patches.h"
 #include "drw.h"
@@ -752,14 +753,14 @@ drw_font_getexts(Fnt *font, const char *text, unsigned int len, unsigned int *w,
 #endif // BAR_PANGO_PATCH
 
 Cur *
-drw_cur_create(Drw *drw, int shape)
+drw_cur_create(Drw *drw, const char *shape)
 {
 	Cur *cur;
 
 	if (!drw || !(cur = ecalloc(1, sizeof(Cur))))
 		return NULL;
 
-	cur->cursor = XCreateFontCursor(drw->dpy, shape);
+	cur->cursor = XcursorLibraryLoadCursor(drw->dpy, shape);
 
 	return cur;
 }
diff --git a/drw.h b/drw.h
index ec24e02..1083436 100644
--- a/drw.h
+++ b/drw.h
@@ -82,7 +82,7 @@ Clr *drw_scm_create(
 );
 
 /* Cursor abstraction */
-Cur *drw_cur_create(Drw *drw, int shape);
+Cur *drw_cur_create(Drw *drw, const char *shape);
 void drw_cur_free(Drw *drw, Cur *cursor);
 
 /* Drawing context manipulation */
diff --git a/dwm.c b/dwm.c
index 6f94b7a..361755e 100644
--- a/dwm.c
+++ b/dwm.c
@@ -30,7 +30,6 @@
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/wait.h>
-#include <X11/cursorfont.h>
 #include <X11/keysym.h>
 #include <X11/Xatom.h>
 #include <X11/Xlib.h>
@@ -40,6 +39,7 @@
 #include <X11/extensions/Xinerama.h>
 #endif /* XINERAMA */
 #include <X11/Xft/Xft.h>
+#include <X11/Xcursor/Xcursor.h>
 
 #include "patches.h"
 #include "drw.h"
@@ -1954,7 +1954,7 @@ void
 drawbar(Monitor *m)
 {
 	Bar *bar;
-	
+
 	#if !BAR_FLEXWINTITLE_PATCH
 	if (m->showbar)
 	#endif // BAR_FLEXWINTITLE_PATCH
@@ -4015,8 +4015,8 @@ setup(void)
 	motifatom = XInternAtom(dpy, "_MOTIF_WM_HINTS", False);
 	#endif // DECORATION_HINTS_PATCH
 	/* init cursors */
-	cursor[CurNormal] = drw_cur_create(drw, XC_left_ptr);
-	cursor[CurResize] = drw_cur_create(drw, XC_sizing);
+	cursor[CurNormal] = drw_cur_create(drw, "left_ptr");
+	cursor[CurResize] = drw_cur_create(drw, "se-resize");
 	#if RESIZEPOINT_PATCH || RESIZECORNERS_PATCH
 	cursor[CurResizeBR] = drw_cur_create(drw, XC_bottom_right_corner);
 	cursor[CurResizeBL] = drw_cur_create(drw, XC_bottom_left_corner);
@@ -4033,7 +4033,7 @@ setup(void)
 	#if DRAGFACT_PATCH && CFACTS_PATCH
 	cursor[CurDragFact] = drw_cur_create(drw, XC_rightbutton);
 	#endif // DRAGFACT_PATCH
-	cursor[CurMove] = drw_cur_create(drw, XC_fleur);
+	cursor[CurMove] = drw_cur_create(drw, "fleur");
 	/* init appearance */
 	#if BAR_VTCOLORS_PATCH
 	get_vt_colors();
-- 
2.51.2

